File: Projects_2013VS/FindPathBetweenSensors_Google/FindPathBetweenSensors_Google.cpp
// FindPathBetweenSensors_Google.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"

#include <array>
#include <limits>
#include <set>
#include <vector>

#include <cassert>
#include <cmath>

struct Location
{
    Location()
        : x(-1.0)
        , y(-1.0)
    {}
    Location(double x_, double y_)
        : x(x_)
        , y(y_)
    {}
    double x;
    double y;

    static double GetDistance(const Location &l1, const Location &l2) {
        const double deltaX = l1.x - l2.x;
        const double deltaY = l1.y - l2.y;
        return sqrt(deltaX*deltaX + deltaY*deltaY);
    }

    bool operator==(const Location &rhs) const {
        return x == rhs.x && y == rhs.y;
    }
};

struct EdgeCross
{
    EdgeCross()
    {}

    EdgeCross(double lb, double ub)
        : lowerBound(lb)
        , upperBound(ub)
    {}

    inline void SetBounds(double lb, double ub)
    {
        lowerBound = lb;
        upperBound = ub;
    }

    inline void SetLowerBound(double lb)
    {
        lowerBound = lb;
    }

    inline void SetUpperBound(double ub)
    {
        upperBound = ub;
    }
    bool IsValid() const {
        return lowerBound != std::numeric_limits<double>::max() &&
               upperBound != std::numeric_limits<double>::min();
    }

    double lowerBound = std::numeric_limits<double>::max();
    double upperBound = std::numeric_limits<double>::min();

    void MergeBounds(const EdgeCross &rhs)
    {
        SetBounds(std::min(lowerBound, rhs.lowerBound),
                  std::max(upperBound, rhs.upperBound));
    }
    enum EDGE
    {
        LEFT = 0UL,
        TOP,
        RIGHT,
        BOTTOM,
        MAX
    };

    using EdgeCrosses = std::array<EdgeCross, EDGE::MAX>;
};

struct EdgeCrossCmpLess
{
    bool operator() (const EdgeCross *lhs, const EdgeCross *rhs) const {
        return lhs->upperBound < rhs->lowerBound;
    }
};

struct Sensor
{
    Sensor()
    {}

    Sensor(const Location &cnter, double rds)
        : center(cnter)
        , radius(rds)
    {
        assert(radius >= 0.0);
    }
    Location center;
    double radius = 0.0;

    bool Crossed(const Sensor &rhs) const {
        double dist = Location::GetDistance(this->center, rhs.center);
        return dist <= fabs(this->radius + rhs.radius);
    }

    bool CrossedWithHLine(const double Y, EdgeCross& ec) const {
        if (Y >= (center.y - radius) && Y <= (center.y + radius)) {
            const double deltaY = center.y - Y;
            const double deltaX = sqrt(radius*radius - deltaY*deltaY);
            ec.SetBounds(center.x - deltaX, center.x + deltaX);
            return true;
        }

        return false;
    }

    bool CrossedWithVLine(const double X, EdgeCross& ec) const {
        if (X >= (center.x - radius) && X <= (center.x + radius)) {
            const double deltaX = center.x - X;
            const double deltaY = sqrt(radius*radius - deltaX*deltaX);
            ec.SetBounds(center.y - deltaY, center.y + deltaY);
            return true;
        }

        return false;
    }
};

struct SensorRange
{
    SensorRange()
    {}

    EdgeCross::EdgeCrosses edgeCrosses;
    std::vector<const Sensor *> sensors;

    void MergeEdgeCross(const EdgeCross::EdgeCrosses &ecs) {
        for (size_t idx = 0; idx < ecs.size(); ++idx) {
            edgeCrosses[idx].MergeBounds(ecs[idx]);
        }
    }
    void Merge(const SensorRange &sr) {
        MergeEdgeCross(sr.edgeCrosses);
        sensors.insert(sensors.end(), sr.sensors.begin(), sr.sensors.end());
    }
};

struct Rectangle {
    Rectangle()
    {}

    Rectangle(const double l, const double r, const double t, const double b)
    {
        assert(l < r);
        assert(b < t);
        edges[EdgeCross::LEFT] = l;
        edges[EdgeCross::RIGHT] = r;
        edges[EdgeCross::TOP] = t;
        edges[EdgeCross::BOTTOM] = b;
    }

    std::array<double, EdgeCross::MAX> edges;
};

void FindCrossBetweenSensorAndRect(const Sensor &sensor,
                                   const Rectangle &rect,
                                   EdgeCross::EdgeCrosses &edgeCrosses)
{
    for (size_t edge = EdgeCross::LEFT; edge < EdgeCross::MAX; ++edge) {
        if (edge & 1UL) {
            sensor.CrossedWithHLine(rect.edges[edge], edgeCrosses[edge]);
        }
        else {
            sensor.CrossedWithVLine(rect.edges[edge], edgeCrosses[edge]);
        }
    }
}

bool GenerateSenorRangeMap(const std::vector<Sensor> &sensors,
                           const Rectangle &rect,
                           const bool breakIfNoPath,
                           const bool vPath,
                           std::vector<SensorRange> &sensorRanges)
{
    auto itEnd = sensors.cend();
    std::vector<size_t> toMerge;
    for (auto it = sensors.cbegin(); it != itEnd; ++it) {
		const size_t CUR_SIZE = sensorRanges.size();
        toMerge.clear();
        toMerge.reserve(CUR_SIZE);
        for (size_t idx = 0; idx < CUR_SIZE; ++idx) {
            auto&& sRange = sensorRanges[idx];
            auto itSenEnd = sRange.sensors.cend();
            for (auto itSen = sRange.sensors.cbegin(); itSen != itSenEnd; ++itSen) {
                if ((*itSen)->Crossed(*it)) {
                    toMerge.push_back(idx);
                    break;
                }
            }
        }

        const SensorRange *workingSR = NULL;
        if (toMerge.empty()) {
            // create a new sensor range
            SensorRange sr;
            sr.sensors.emplace_back(&(*it));
            FindCrossBetweenSensorAndRect(*it, rect, sr.edgeCrosses);
            sensorRanges.push_back(sr);
            workingSR = &sensorRanges.back();
        }
        else {
            // merge all existing sensor range
            // 1. merge the rest into the first senror range
            // 2. erase the sensor ranges except the very first
            auto itMergeREnd = toMerge.crend() - 1;
            for (auto itMergeR = toMerge.crbegin(); itMergeR != itMergeREnd; ++itMergeR) {
                auto&& sRangeToKeep = sensorRanges[*toMerge.begin()];
                auto&& sRangeToMerge = sensorRanges[*itMergeR];
                sRangeToKeep.Merge(sRangeToMerge);
                sensorRanges.erase(sensorRanges.begin() + *itMergeR);
            }
            // merege with the sensor
            EdgeCross::EdgeCrosses edgeCrosses;
            FindCrossBetweenSensorAndRect(*it, rect, edgeCrosses);
            auto&& sRangeToKeep = sensorRanges[*toMerge.begin()];
            sRangeToKeep.MergeEdgeCross(edgeCrosses);
            sRangeToKeep.sensors.push_back(&(*it));
            workingSR = &sRangeToKeep;
        }
        if (workingSR && breakIfNoPath) {
            if ((vPath &&
                 workingSR->edgeCrosses[EdgeCross::LEFT].IsValid() &&
                 workingSR->edgeCrosses[EdgeCross::RIGHT].IsValid()) ||
                (!vPath &&
                 workingSR->edgeCrosses[EdgeCross::TOP].IsValid() &&
                 workingSR->edgeCrosses[EdgeCross::BOTTOM].IsValid())) {
                    return false;
            }
        }
    }
    return true;
}

bool ProcessCornerCase(const bool vPath,
                       const EdgeCross::EdgeCrosses &edgeCrosses,
                       const Rectangle &rect,
                       EdgeCross &edge)
{
    if (edge.IsValid()) {
        if (vPath) {
            // left corner
            if (edgeCrosses[EdgeCross::LEFT].IsValid()) {
                edge.SetLowerBound(rect.edges[EdgeCross::LEFT]);
            }
            // right corner
            if (edgeCrosses[EdgeCross::RIGHT].IsValid()) {
                edge.SetUpperBound(rect.edges[EdgeCross::RIGHT]);
            }
        }
        else {
            // top corner
            if (edgeCrosses[EdgeCross::TOP].IsValid()) {
                edge.SetUpperBound(rect.edges[EdgeCross::TOP]);
            }
            // bottom corner
            if (edgeCrosses[EdgeCross::BOTTOM].IsValid()) {
                edge.SetLowerBound(rect.edges[EdgeCross::BOTTOM]);
            }
        }
        return true;
    }

    return false;
}

using Path = std::array<Location, 2>;
using EdgeCrossSet = std::set<const EdgeCross*, EdgeCrossCmpLess>;

bool GetMidPointOfFirstGap(const EdgeCrossSet &ecs,
                           const Rectangle &rect,
                           const bool vPath,
                           double &val)
{
    const double lowerBound = vPath ?
        rect.edges[EdgeCross::LEFT] : rect.edges[EdgeCross::BOTTOM];
    const double upperBound = vPath ?
        rect.edges[EdgeCross::RIGHT] : rect.edges[EdgeCross::TOP];
    if (ecs.empty()) {
        val = (lowerBound + upperBound) / 2;
        return true;
    }
    auto itFirst = ecs.cbegin();
    if ((*itFirst)->lowerBound > lowerBound) {
        val = (lowerBound + (*itFirst)->lowerBound) / 2;
        return true;
    }
    auto itSecond = ++ecs.cbegin();
    if (itSecond == ecs.cend()) {
        if ((*itFirst)->upperBound < upperBound) {
            val = ((*itFirst)->upperBound + upperBound) / 2;
            return true;
        }
        return false;
    }

    if ((*itFirst)->upperBound < (*itSecond)->lowerBound) {
        val = ((*itFirst)->upperBound + (*itSecond)->lowerBound) / 2;
        return true;
    }
    return false;
}

bool FindPath(const std::vector<Sensor> &sensors,
              const Rectangle &rect,
              const bool vPath,
              Path &path)
{
    std::vector<SensorRange> sensorRanges;
    if (!sensors.empty()) {
        if (!GenerateSenorRangeMap(sensors, rect, true, vPath, sensorRanges)) {
            return false;
        }
    }

    if (sensorRanges.empty()) {
        if (vPath) {
            path[0] = Location{ rect.edges[EdgeCross::LEFT], rect.edges[EdgeCross::TOP] };
            path[1] = Location{ rect.edges[EdgeCross::LEFT], rect.edges[EdgeCross::BOTTOM] };
        }
        else {
            path[0] = Location{ rect.edges[EdgeCross::LEFT], rect.edges[EdgeCross::TOP] };
            path[1] = Location{ rect.edges[EdgeCross::RIGHT], rect.edges[EdgeCross::TOP] };
        }
        return true;
    }

    EdgeCrossSet ecsFrom;
    EdgeCrossSet ecsTo;
    auto itEnd = sensorRanges.end();
    for (auto it = sensorRanges.begin(); it != itEnd; ++it) {
        if (vPath && ProcessCornerCase(vPath,
                                       it->edgeCrosses,
                                       rect, 
                                       it->edgeCrosses[EdgeCross::TOP])) {
            ecsFrom.insert(&(it->edgeCrosses[EdgeCross::TOP]));
        }
        if (vPath && ProcessCornerCase(vPath,
                                       it->edgeCrosses,
                                       rect,
                                       it->edgeCrosses[EdgeCross::BOTTOM])) {
            ecsTo.insert(&(it->edgeCrosses[EdgeCross::BOTTOM]));
        }
        if (!vPath && ProcessCornerCase(vPath,
                                        it->edgeCrosses,
                                        rect,
                                        it->edgeCrosses[EdgeCross::LEFT])) {
            ecsFrom.insert(&(it->edgeCrosses[EdgeCross::LEFT]));
        }
        if (!vPath && ProcessCornerCase(vPath,
                                        it->edgeCrosses,
                                        rect,
                                        it->edgeCrosses[EdgeCross::RIGHT])) {
            ecsTo.insert(&(it->edgeCrosses[EdgeCross::RIGHT]));
        }
    }

    double valFrom, valTo;
    if (GetMidPointOfFirstGap(ecsFrom, rect, vPath, valFrom) &&
        GetMidPointOfFirstGap(ecsTo, rect, vPath, valTo)) {
        path[0] = vPath ? Location(valFrom, rect.edges[EdgeCross::TOP]) :
                          Location(rect.edges[EdgeCross::LEFT], valFrom);
        path[1] = vPath ? Location(valTo, rect.edges[EdgeCross::BOTTOM]) :
                          Location(rect.edges[EdgeCross::RIGHT], valTo);

        return true;
    }

    return false;
}

void TestFindPathBetweenSensors()
{
    Path path;
    const Rectangle rect = { 1.0, 9.0, 8.0, 1.0 };
    {
        const std::vector<Sensor> sensors{ { { 2.0, 3.0 }, 1.0 },
                                           { { 4.0, 3.0 }, 1.0 },
                                           { { 6.0, 3.0 }, 1.0 },
                                           { { 8.0, 3.0 }, 1.0 }
                                         };
        assert(FindPath(sensors, rect, true, path) == false);
        assert(FindPath(sensors, rect, false, path) == true);
        assert(path.at(0) == Location(1.0, 2.0) && path.at(1) == Location(9.0, 2.0));
    }

    {
        const std::vector<Sensor> sensors{ { { 2.0, 3.0 }, 1.0 },
                                           { { 4.0, 3.0 }, 1.0 },
                                           { { 6.0, 3.0 }, 1.0 },
                                           { { 8.0, 3.0 }, 1.0 },
                                           { { 8.0, 8.0 }, 1.0 },
                                           { { 7.0, 7.0 }, 1.0 }
                                         };
        assert(FindPath(sensors, rect, true, path) == false);
        assert(FindPath(sensors, rect, false, path) == true);
        assert(path.at(0) == Location(1.0, 2.0) && path.at(1) == Location(9.0, 2.0));
    }

    {
        const std::vector<Sensor> sensors{ { { 2.0, 7.2 }, 1.0 },
                                           { { 3.0, 6.0 }, 1.0 },
                                           { { 4.0, 4.5 }, 1.0 },
                                           { { 8.0, 7.2 }, 1.0 },
                                           { { 7.0, 6.2 }, 1.0 },
                                           { { 2.0, 3.0 }, 1.0 },
                                           { { 3.0, 4.0 }, 1.0 },
                                           { { 9.0, 1.0 }, 1.0 }
                                         };
        assert(FindPath(sensors, rect, true, path) == true);
        assert(path.at(0) == Location(5.0, 8.0) && path.at(1) == Location(4.5, 1.0));
        assert(FindPath(sensors, rect, false, path) == true);
        assert(path.at(0) == Location(1.0, 2.0) && path.at(1) == Location(9.0, 4.6));
    }
}

int _tmain(int argc, _TCHAR* argv[])
{
    TestFindPathBetweenSensors();
    return 0;
}

File: PetersPyProjects/PetersPyProjects/FindPathBetweenSensors_Google/FindPathBetweenSensors_Google.py
from EdgeCross import EdgeCross
from EdgeCross import EdgeCrosses
from Location import Location
from Rectangle import Rectangle
from Sensor import Sensor
from SensorRange import SensorRange

import sortedcontainers
from exceptions import AttributeError
import sys

class FindPathBetweenSensors_Google(object):
    """FindPathBetweenSensors of class"""

    def __init__(self, *args, **kwargs):
        return super(FindPathBetweenSensors_Google, self).__init__(*args, **kwargs)

    @staticmethod
    def __FindCrossBetweenSensorAndRect(sensor, rect, edgeCrosses):
        for edge in range(0, EdgeCross.MAX()) :
            if edge & 1:
                sensor.CrossedWithHLine(rect.edges[edge], edgeCrosses.ecs[edge])
            else:
                sensor.CrossedWithVLine(rect.edges[edge], edgeCrosses.ecs[edge])

    @staticmethod
    def __GenerateSensorRangeMap(sensors, rect, breakIfNoPath, vPath, sensorRanges):
        for sensor in sensors:
            toMerge = []
            CUR_SIZE = len(sensorRanges)
            for idx in range(0, CUR_SIZE):
                for s in sensorRanges[idx].sensors:
                    if s.Crossed(sensor) :
                        toMerge.append(idx)
                        break
            
            workingSR = None
            if len(toMerge) == 0:
                # create a new sensor range
                sr = SensorRange()
                sr.sensors.append(sensor)
                FindPathBetweenSensors_Google.__FindCrossBetweenSensorAndRect(sensor, rect, sr.edgeCrosses)
                sensorRanges.append(sr)
                workingSR = sr
            else:
                # merge all existing sensor range
                # 1. merge the rest into the first senror range
                # 2. erase the sensor ranges except the very first
                for idx in range(len(toMerge) - 1, 0, -1):
                    sRangeToKeep = sensorRanges[toMerge[0]]
                    sRangeToMerge = sensorRanges[toMerge[idx]]
                    sRangeToKeep.Merge(sRangeToMerge)
                    del sensorRanges[toMerge[idx]]
                
                # merge with the sensor
                edgeCrosses = EdgeCrosses()
                FindPathBetweenSensors_Google.__FindCrossBetweenSensorAndRect(sensor, rect, edgeCrosses)
                sRangeToKeep= sensorRanges[toMerge[0]]
                sRangeToKeep.MergeEdgeCross(edgeCrosses)
                sRangeToKeep.sensors.append(sensor)
                workingSR = sRangeToKeep
            
            if workingSR is not None and breakIfNoPath:
                if (vPath and workingSR.edgeCrosses.ecs[EdgeCross.LEFT()].IsValid() and workingSR.edgeCrosses.ecs[EdgeCross.RIGHT()].IsValid()) or \
                   (not vPath and workingSR.edgeCrosses.ecs[EdgeCross.TOP()].IsValid() and workingSR.edgeCrosses.ecs[EdgeCross.BOTTOM()].IsValid()):
                    return False
        return True
       
    @staticmethod
    def __ProcessCornerCase(vPath, edgeCrosses, rect, edge):
        if edge.IsValid():
            if vPath:
                # left corner
                if edgeCrosses.ecs[EdgeCross.LEFT()].IsValid():
                    edge.SetLowerBound(rect.edges[EdgeCross.LEFT()])
                # right corner
                if edgeCrosses.ecs[EdgeCross.RIGHT()].IsValid():
                    edge.SetUpperBound(rect.edges[EdgeCross.RIGHT()])
            else:
                # top corner
                if edgeCrosses.ecs[EdgeCross.TOP()].IsValid():
                    edge.SetUpperBound(rect.edges[EdgeCross.TOP()])
                # bottom corner
                if edgeCrosses.ecs[EdgeCross.BOTTOM()].IsValid():
                    edge.SetLowerBound(rect.edges[EdgeCross.BOTTOM()])
            return True
        return False

    @staticmethod
    def __GetMidPointOfFirstGap(ecs, rect, vPath, val):
        lowerBound = rect.edges[EdgeCross.BOTTOM()]
        upperBound = rect.edges[EdgeCross.TOP()]
        if vPath:
            lowerBound = rect.edges[EdgeCross.LEFT()]
            upperBound = rect.edges[EdgeCross.RIGHT()]
        
        if len(ecs) == 0:
            val[0] = (lowerBound + upperBound)/2
            return True
        
        if ecs[0].lowerBound > lowerBound:
            val[0] = (ecs[0].lowerBound + lowerBound)/2
            return True

        if len(ecs) == 1:
            if ecs[0].upperBound < upperBOund:
                val[0] = (ecs[0].upperBound + upperBound)/2
                return True
            return False

        if ecs[0].upperBound < ecs[1].lowerBound:
            val[0] = (ecs[0].upperBound + ecs[1].lowerBound)/2
            return True
        
        return False

    @staticmethod
    def FindPath(sensors, rect, vPath, path):
        try:
            sensorRanges = list()
            if len(sensors) != 0:
                if not FindPathBetweenSensors_Google.__GenerateSensorRangeMap(sensors, rect, True, vPath, sensorRanges):
                    return False
            if len(sensorRanges) == 0:
                if vPath:
                    path.append(Location(rect.edges[EdgeCross.LEFT()], rect.edges[EdgeCross.TOP()]))
                    path.append(Location(rect.edges[EdgeCross.LEFT()], rect.edges[EdgeCross.BOTTOM()]))
                else:
                    path.append(Location(rect.edges[EdgeCross.LEFT()], rect.edges[EdgeCross.TOP()]))
                    path.append(Location(rect.edges[EdgeCross.RIGHT()], rect.edges[EdgeCross.TOP()]))
                return True

            ecsFrom = sortedcontainers.SortedSet([])
            ecsTo = sortedcontainers.SortedSet([])
            for sr in sensorRanges:
                if vPath and FindPathBetweenSensors_Google.__ProcessCornerCase(vPath, sr.edgeCrosses, rect, sr.edgeCrosses.ecs[EdgeCross.TOP()]):
                    ecsFrom.add(sr.edgeCrosses.ecs[EdgeCross.TOP()])
                if vPath and FindPathBetweenSensors_Google.__ProcessCornerCase(vPath, sr.edgeCrosses, rect, sr.edgeCrosses.ecs[EdgeCross.BOTTOM()]):
                    ecsTo.add(sr.edgeCrosses.ecs[EdgeCross.BOTTOM()])
                if (not vPath) and FindPathBetweenSensors_Google.__ProcessCornerCase(vPath, sr.edgeCrosses, rect, sr.edgeCrosses.ecs[EdgeCross.LEFT()]):
                    ecsFrom.add(sr.edgeCrosses.ecs[EdgeCross.LEFT()])
                if (not vPath) and FindPathBetweenSensors_Google.__ProcessCornerCase(vPath, sr.edgeCrosses, rect, sr.edgeCrosses.ecs[EdgeCross.RIGHT()]):
                    ecsTo.add(sr.edgeCrosses.ecs[EdgeCross.RIGHT()])
      
            valFrom = [0.0]
            valTo = [0.0]
            if FindPathBetweenSensors_Google.__GetMidPointOfFirstGap(ecsFrom, rect, vPath, valFrom) and \
               FindPathBetweenSensors_Google.__GetMidPointOfFirstGap(ecsTo, rect, vPath, valTo) :
                if vPath:
                    path.append(Location(valFrom[0], rect.edges[EdgeCross.TOP()]))
                    path.append(Location(valTo[0], rect.edges[EdgeCross.BOTTOM()]))
                else:
                    path.append(Location(rect.edges[EdgeCross.LEFT()], valFrom[0]))
                    path.append(Location(rect.edges[EdgeCross.RIGHT()], valTo[0]))
                return True
        except AttributeError() as e:
            print ("AttributeError: {0}".format(e))
        except:
            print ("Unexpected error: ", sys.exc_info()[0])
        return False
        

# TEST
def TestFindPathBetweenSensors():
    path = []
    rect = Rectangle(1.0, 9.0, 8.0, 1.0)
    sensors = [ Sensor(Location(2.0, 3.0), 1.0), \
                Sensor(Location(4.0, 3.0), 1.0), \
                Sensor(Location(6.0, 3.0), 1.0), \
                Sensor(Location(8.0, 3.0), 1.0) ]
    assert(FindPathBetweenSensors_Google.FindPath(sensors, rect, True, path) == False)
    assert(FindPathBetweenSensors_Google.FindPath(sensors, rect, False, path) == True)
    assert(path[0] == Location(1.0, 2.0))
    assert(path[1] == Location(9.0, 2.0))

    path = []
    sensors = [ Sensor(Location(2.0, 3.0), 1.0), \
                Sensor(Location(4.0, 3.0), 1.0), \
                Sensor(Location(6.0, 3.0), 1.0), \
                Sensor(Location(8.0, 3.0), 1.0), \
                Sensor(Location(8.0, 8.0), 1.0), \
                Sensor(Location(7.0, 7.0), 1.0) ]
    assert(FindPathBetweenSensors_Google.FindPath(sensors, rect, True, path) == False)
    assert(FindPathBetweenSensors_Google.FindPath(sensors, rect, False, path) == True)
    assert(path[0] == Location(1.0, 2.0))
    assert(path[1] == Location(9.0, 2.0))

    path = []
    sensors = [ Sensor(Location(2.0, 7.2), 1.0), \
                Sensor(Location(3.0, 6.0), 1.0), \
                Sensor(Location(4.0, 4.5), 1.0), \
                Sensor(Location(8.0, 7.2), 1.0), \
                Sensor(Location(7.0, 6.2), 1.0), \
                Sensor(Location(2.0, 3.0), 1.0), \
                Sensor(Location(3.0, 4.0), 1.0), \
                Sensor(Location(9.0, 1.0), 1.0) ]
    assert(FindPathBetweenSensors_Google.FindPath(sensors, rect, True, path) == True)
    assert(path[0] == Location(5.0, 8.0))
    assert(path[1] == Location(4.5, 1.0))
    path = []
    assert(FindPathBetweenSensors_Google.FindPath(sensors, rect, False, path) == True)
    assert(path[0] == Location(1.0, 2.0))
    assert(path[1] == Location(9.0, 4.6))

def Main():
    TestFindPathBetweenSensors()
    
if __name__ == '__main__':
    Main()
